cirq/value/product_state.py

```diff
@@ -0,0 +1,319 @@
+                             "with other TensorProductStates.")
+
+        dupe_qubits = set(other.states.keys()) & set(self.states.keys())
+        if len(dupe_qubits) != 0:
+            raise ValueError(
+                "You tried to tensor two states, "
+                "but both contain factors for these qubits: {}".format(
+                    sorted(dupe_qubits)))
+
+        new_states = self.states
+        new_states.update(other.states)
+        return ProductState(new_states)
+
+    def __str__(self) -> str:
+        return ' * '.join(f'{st}({q})' for q, st in self.states.items())
+
+    def __repr__(self) -> str:
+        states_dict_repr = ', '.join(
+            f'{repr(key)}: {repr(val)}' for key, val in self.states.items())
+        return 'cirq.ProductState({%s})' % states_dict_repr
+
+    def __getitem__(self, qubit: 'cirq.Qid') -> _NamedOneQubitState:
+        """Return the _NamedOneQubitState at the given qubit."""
+        return self.states[qubit]
+
+    def __iter__(self) -> Iterable[Tuple['cirq.Qid', _NamedOneQubitState]]:
+        yield from self.states.items()
+
+    def __len__(self) -> int:
+        return len(self.states)
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, ProductState):
+            return False
+
+        return self.states == other.states
+
+    def __hash__(self):
+        return hash(tuple(self.states.items()))
+
+    def _json_dict_(self):
+        return {
+            'cirq_type': self.__class__.__name__,
+            'states': list(self.states.items()),
+        }
+
+    @classmethod
+    def _from_json_dict_(cls, states, **kwargs):
+        return cls(states=dict(states))
+
+    def state_vector(self, qubit_order: 'cirq.QubitOrder' = None) -> np.ndarray:
+        """The state-vector representation of this state."""
+        from cirq import ops
+        if qubit_order is None:
+            qubit_order = ops.QubitOrder.DEFAULT
+        qubit_order = ops.QubitOrder.as_qubit_order(qubit_order)
+        qubits = qubit_order.order_for(self.qubits)
+
+        mat = 1.0 + 0.j
+        for qubit in qubits:
+            oneq_state = self[qubit]
+            state_vector = oneq_state.state_vector()
+            mat = np.kron(mat, state_vector)
+
+        return mat
+
+    def projector(self, qubit_order: 'cirq.QubitOrder' = None) -> np.ndarray:
+        """The projector associated with this state expressed as a matrix.
+
+        This is |s⟩⟨s| where |s⟩ is this state.
+        """
+        from cirq import ops
+        if qubit_order is None:
+            qubit_order = ops.QubitOrder.DEFAULT
+        qubit_order = ops.QubitOrder.as_qubit_order(qubit_order)
+        qubits = qubit_order.order_for(self.qubits)
+
+        mat = 1.0 + 0.j
+        for qubit in qubits:
+            oneq_state = self[qubit]
+            oneq_proj = oneq_state.projector()
+            mat = np.kron(mat, oneq_proj)
+        return mat
+
+
+class _PauliEigenState(_NamedOneQubitState):
```

Comment:
For future extensibility: does this need to have "Pauli" in the name? AFAICT, the implementation is appropriate more generally (modulo a single type annotation below) .

Reply:
There is dependency in str() and init() on having only +1/-1 eigenvalues available for this state. This is specific to involutory matrices - but in this context I think it's fine to call it PauliEigenState - if we need a more generic EigenState, we can extract it. I would like to push this in as I think it is in a good enough shape to be iterated on if needed. @viathor do you have any big blockers still outstanding?

Reply:
Good point about nature of the eigenvalues. No other blockers.